pyomo/core/plugins/transform/equality_propagate.py

```diff
@@ -0,0 +1,221 @@
+from pyomo.util.plugin import alias
+from six import iteritems
+
+__author__ = "Qi Chen <qichen at andrew.cmu.edu>"
+
+
+def _build_equality_set(m, var_to_id, id_to_var):
+    """Construct an equality set map.
+
+    Maps all variables to the set of variables that are linked to them by
+    equality. Mapping takes place using ComponentUID. That is, if you have x =
+    y, then you would have xUID -> set([xUID, yUID]) and yUID -> set([xUID,
+    yUID]) in the mapping.
+
+    var_to_id and id_to_var are the model symbol map and reverse map,
+    respectively.
+
+    """
+    #: dict: map of var UID to the set of all equality-linked var UIDs
+    eq_var_map = {}
+    for constr in m.component_data_objects(ctype=Constraint,
+                                           active=True,
+                                           descend_into=True):
+        # Check to make sure the constraint is of form v1 - v2 == 0
+        if constr.lower == 0 and constr.upper == 0 \
+                and isinstance(constr.body, _SumExpression) \
+                and len(constr.body._args) == 2 \
+                and 1 in constr.body._coef \
+                and -1 in constr.body._coef \
+                and isinstance(constr.body._args[0], _GeneralVarData) \
+                and isinstance(constr.body._args[1], _GeneralVarData) \
+                and constr.body._const == 0:
+            v1 = constr.body._args[0]
+            v2 = constr.body._args[1]
+            set1 = eq_var_map.get(var_to_id[v1], set([var_to_id[v1]]))
+            set2 = eq_var_map.get(var_to_id[v2], set([var_to_id[v2]]))
+            union = set1.union(set2)
+            for vID in union:
+                eq_var_map[vID] = union
+
+    return eq_var_map
+
+
+class FixedVarPropagator(IsomorphicTransformation):
+    """Propagates variable fixing for equalities of type x = y.
+
+    If x is fixed and y is not fixed, then this transformation will fix y to
+    the value of x.
+
+    This transformation can also be performed as a temporary transformation,
+    whereby the transformed variables are saved and can be later unfixed.
+
+    """
+
+    alias('core.propagate_fixed_vars', doc=__doc__)
+
+    def __init__(self):
+        """Initialize the transformation."""
+        super(FixedVarPropagator, self).__init__()
+        self._tmp_propagate_fixed = set()
+        self._transformed_instance = None
+
+    def _apply_to(self, instance, tmp=False):
+        """Apply the transformation.
+
+        Args:
+            instance (Block): the block on which to search for x == y
+                constraints. Note that variables may be located anywhere in
+                the model.
+            tmp (bool, optional): Whether the variable fixing will be temporary
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: if two fixed variables x = y have different values.
+
+        """
+        if tmp:
+            self._transformed_instance = instance
```

Comment:
What happens if you do the same transformation twice at different times and want to revert one? This isn't currently something that I'm worried about, so I'll go ahead and stick the the state information on the instance, but it's something to consider.